Added coding standards to the project. (#139)
keycloak/keycloak-admin-ui: CODING_GUIDELINES.md

# Coding Guidelines

## Typescript

> The Keycloak UI projects uses best practices based off the offical [React TypeScript Cheatsheet](https://react-typescript-cheatsheet.netlify.app/), with modificiations for this project. The React TypeScript Cheatsheet is maintained and used by developers through out the world, and is a place where developers can bring togehter lessons learned using TypeScript and React.

Reviewer: typo - "modifications"

Author: Done

> The Keycloak UI projects uses best practices based off the offical [React TypeScript Cheatsheet](https://react-typescript-cheatsheet.netlify.app/), with modificiations for this project. The React TypeScript Cheatsheet is maintained and used by developers through out the world, and is a place where developers can bring togehter lessons learned using TypeScript and React.

Reviewer 2: offical - official

Author: done

### Imports

Since we are using TypeScript 4.x + for this project, default imports should conform to the new standard set forth in TypeScript 2.7:

```javascript

> import React from 'react';

Author: Done

> import ReactDOM from 'react-dom';

Author: Done

```

For imports that are not the default import use the following syntax:

```javascript

> import { X1, X2, ... Xn } from 'package-x';

Author: Done

```

### Props

For props we are using **type** instead of **interfaces**. The reason to use types instead of interfaces is for consistency betweeen the views and because it's more constrained (See [Types or Interfaces](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces) for more clarifiation). By using types we are ensuring that we will not deviate from the agreed upon [contract](https://dev.to/reyronald/typescript-types-or-interfaces-for-react-component-props-1408).

The following is an example of using a type for props:

```javascript

> import React, { ReactNode } from 'react'

Author: Done

...

export type ExampleComponentProps = {
    message: string;
    children: ReactNode;
}
```

> ### Maintiaing state with types

Reviewer 3: ```suggestion
### Maintaining state with types
```

Author: Done.

> When maintaining state for a component it is recommend that you use a [type instead of an interface](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).

Reviewer 3: ```suggestion
When maintaining state for a component, it is recommended that you use a [type instead of an interface](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).
```

Author: Done.

```javascript
type ApiDrawerState = {
  currentApiId: string,
  isExpanded: boolean,
};
```

> ### Intefaces

Reviewer 3: ```suggestion
### Interfaces
```

Author: Done

Interfaces should be used for all public facing API definitions. A table describing when to use interfaces vs. types can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).

### Function Components

> This project uses function components and hooks over class components. When coding function components in typescript a developer should include any specific props that they need.

Reviewer 3: ```suggestion
This project uses function components and hooks over class components. When coding function components in typescript, a developer should include any specific props that they need.
```

Author: Done

```javascript

> import React, { FunctionComponent } from 'react';

Author: Done

...

export const ExampleComponent: FunctionComponent<ExampleComponentProps> = ({ message, children }: ExampleComponentProps) => (
  <ReactFragment>
    <div>{message}</div>
    <div>{children}</div>
  </<ReactFragment>>
);
```

> For components that do not have any additonal props an empty object should be used instead:

Author: done

```javascript
import React, { FunctionComponent } from 'react';

...

export const ExampleNoPropsComponent: FunctionComponent<{}> = () => (
  <div>Example Component with no props</div>
);
```

> Additonal details around function components can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/function_components).

Reviewer 2: Additional

Author: done

### Hooks

> When using hooks with Typescript there are few recommendations that we follow below. Additional recommendations besides the ones mention in this document can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks).

Reviewer 3: ```suggestion
When using hooks with Typescript there are few recommendations that we follow below. Additional recommendations besides the ones mentioned in this document can be found [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks).
```

### Inference vs Types for useState

> Currently we recommend using inference for the primative types booleans, numbers, and string when using useState. Anything other then these 3 types should use a declartive syntax to specify what is expected. For example the following is an example of how to use inference:

Reviewer 3: ```suggestion
Currently we recommend using inference for primitive types booleans, numbers, and strings when using useState. Anything other then these 3 types should use a declarative syntax to specify what is expected. For example the following is an example of how to use inference:
```

Author: Done.

```javascript
const [isEnabled, setIsEnabled] = React.useState(false);
```

> Here is an example how to use a declaritive syntax. When using a declearative syntax if the value can be null that will also need to be specified:

Reviewer 3: ```suggestion
Here is an example how to use a declarative syntax. When using a declarative syntax, if the value can be null, that will also need to be specified:
```

Author: Done

```javascript
const [user, setUser] = React.useState<IUser | null>(null);

...

setUser(newUser);

```

> #### useReducres

Reviewer 3: ```suggestion
#### useReducer
```

Author: Done

When using reducers make sure you specify the [return type and not use inference](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#usereducer).

#### useEffect

For useEffect only [return the function or undefined](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useeffect).

#### useRef

When using useRef there are two options with Typescript. The first one is when creating a [read-only ref](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref).

```javascript
const refExample = useRef<HTMLElement>(null!);
```

By passing in null! it will prevent Typescript from returning an error saying refExample maybe null.

> The second option is for createing [mutable refs](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/hooks#useref) that you will manage.

Author: done

```javascript
const refExampleMutable = (useRef < HTMLElement) | (null > null);
```

### Additional Typescript Pointers

Besides the details outlined above a list of recommendations for Typescript is maintained by several Typescript React developers [here](https://react-typescript-cheatsheet.netlify.app/). This is a great reference to use for any additonal questions that are not outlined within the coding standards.

> 

Author: done

## CSS

We use custom CSS in rare cases where PatternFly styling does not meet our design guidelines. If styling needs to be added, we should first check that the Patternfly component is being properly built and whether a variant is already provided to meet the use case. Next, Patternfly layouts should be used for most positioning of components. For one-off tweaks (e.g. spacing an icon slightly away from the text next to it), a PatternFly utility class should be used. In all cases, Patternfly variables should be used for colors, spacing, etc. rather than hard coding color or pixel values.

We will use one global CSS file to surface customization variables. Styles particular to a component should be located in a .CSS file within the component’s folder. A modified BEM naming convention should be used as detailed below.

### Location of files, location of classes

> * Global styling should be located…?  */static/resources/style.css*.

Reviewer 4: Global styling should be in `./public/index.css` right?

Author: Yep let me fix it.

* The CSS relating to a single component should be located in a file within each component’s folder.

> * Should we set up SCSS? *Not for now*

Reviewer: Did we decide yes or no for SCSS?

Reviewer 4: I think the answer was we use it when we need to? As we in the first place would like to minimize the css.

Reviewer 5: Right.  Let's hold off until if and when we definitely need it.

Author: Do you want me to remove this line?

Reviewer 5: Yes.  Remove it.

Author: done.

### Naming CSS classes

PatternFly reference https://pf4.patternfly.org/guidelines#variables

For the Keycloak admin console, we modify the PatternFly convention to namespace the classes and variables to the Keycloak packages.

**Class name**
```css
.keycloak-admin--block[__element][--modifier][--state][--breakpoint][--pseudo-element]
```

**Examples of custom CSS classes**
```css
// Modification to all data tables throughout Keycloak admin 
.keycloak-admin--data-table {
...
}

> // Data tables throughout kogito that are marked as compact 

Reviewer: Replace kogito with keycloak

Author: done

.keycloak-admin--data-table--compact {
...
}

// Child elements of a compact data-table
// Don’t increase specificity with a selector like this: 
// .keycloak-admin--data-table--compact .data-table-item
// Instead, follow the pattern for a single class on the child
.keycloak-admin--data-table__data-table-item--compact {
...
}

// Compact data table just in the management console at the lg or higher breakpoint

> .kogito-management-console--data-table--compact--lg {

Reviewer 4: ```suggestion
.keycloak-admin--data-table--compact--lg {
```

     return (
       <>
         <InfoCircleIcon
           className="pf-u-mr-sm" // utility class 
           color="var(--pf-global--info-color--100)"
         />{' '}
         {titleText}{' '}
       </>
     );
   case 'failure':
     return (
       <>
         <InfoCircleIcon
           className="pf-u-mr-sm"
           color="var(--pf-global--danger-color--100)"
         />{' '}
         {titleText}{' '}
       </>
     );
 } 
 ``` 
**Better way with a custom class**
```css
switch (titleStatus) {
   case 'success':
     return (
       <>
         <InfoCircleIcon
           className="keycloak-admin--icon--info" // use a new kogito- class 
         />{' '}
         {titleText}{' '}
       </>
     );
   case 'failure':
     return (
       <>
         <InfoCircleIcon
           className="keycloak-admin--icon--info"
         />{' '}
         {titleText}{' '}
       </>
     );
 }
```

## Resources

* [Patternfly Docs](https://www.patternfly.org/v4/)
* [Katacoda PatternFly tutorials](https://www.patternfly.org/v4/documentation/react/overview/react-training)
* [PatternFly global CSS variables](https://www.patternfly.org/v4/documentation/overview/global-css-variables)

> * [PatternFly CSS utility classes](https://www.patternfly.org/v4/documentation/react/overview/react-training)

Reviewer 2: the css utility classes link looks like a copy/paste error, it goes to the generic training page

Author: fixed

**Custom property**
```css
--keycloak-admin--block[__element][--modifier][--state][--breakpoint][--pseudo-element]--PropertyCamelCase
```

**Example of a CSS custom property**
```css
// Modify the height of the brand image
--keycloak-admin--brand--Height: var(--pf-global--spacer--xl); 
```

**Example**
```css
// Don’t increase specificity
// Don’t use pixel values
.keycloak-admin--manage-columns__modal .pf-c-dropdown {
   margin-bottom: 24px
}

// Do use a new class
// Do use a PatternFly global spacer variable
.keycloak-admin--manage-columns__dropdown {
   margin-bottom: var(--pf-global--spacer--xl);
}
```
### Using utility classes

Utility classes can be used to add specific styling to a component, such as margin-bottom or padding. However, their use should be limited to one-off styling needs.  

For example, instead of using the utility class for margin-right multiple times, we should define a new Keycloak admin console class that adds *margin-right: var(--pf-global--spacer--sm);*
And in this example, the new class can set the color appropriately as well.

> **Using a utility class **

Reviewer: Extra space here, too, that I think is messing up the markdown. Then that's it, I swear! :)

Author: lol thanks sarah!

Author: done.

```

For imports that are not the default import use the following syntax:

```javascript
import { X1, X2, ... Xn } from "package-x";
```

### Props

For props we are using **type** instead of **interfaces**. The reason to use types instead of interfaces is for consistency between the views and because it"s more constrained (See [Types or Interfaces](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces) for more clarification). By using types we are ensuring that we will not deviate from the agreed upon [contract](https://dev.to/reyronald/typescript-types-or-interfaces-for-react-component-props-1408).

The following is an example of using a type for props:

```javascript
import React, { ReactNode } from "react"

...

export type ExampleComponentProps = {
    message: string;
    children: ReactNode;
}
```

### Maintaining state with types

When maintaining state for a component, it is recommended that you use a [type instead of an interface](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces).

```javascript
type ApiDrawerState = {
  currentApiId: string,
  isExpanded: boolean,
};
```

> 

Reviewer 5: I don't think this makes sense.  Using `type` instead of `interface` is not about state management.  

Since `type instead of an interface` is already in the cheat sheet, I think we can get rid of this section altogether.

Author: It was in this section [here](https://react-typescript-cheatsheet.netlify.app/docs/basic/getting-started/types_or_interfaces) if you need to maintain state of a component with an object, you should try to use type instead of an interface.  I can try rewording this.

### State management

We have made a conscious decision to stay away from state management technologies such as Redux.  These overarching state management schemes tend to be overly complex and encourage dumping everything into the global state.  
 
Instead, we are following a simple philosophy that state should remain close to where it is used and moved to a wider scope only as truly needed.  This encourages encapsulation and makes management of the state much simpler.
 
The way this plays out in our application is that we first prefer state to remain in the scope of the component that uses it.  If the state is required by more than one component, we move to a more complex strategy for management of that state.  In other words, in order of preference, state should be managed by:
* Storing in the component that uses it.
* If #1 is not sufficient, [lift state up](https://reactjs.org/docs/lifting-state-up.html).
* If #2 is not sufficient, try [component composition](https://reactjs.org/docs/context.html#before-you-use-context).

> * If #3, is not sufficient, use a [global context](https://reactjs.org/docs/context.html).

Author: Got it thanks.

**Custom property**
```css
--keycloak-admin--block[__element][--modifier][--state][--breakpoint][--pseudo-element]--PropertyCamelCase
```

**Example of a CSS custom property**
```css
// Modify the height of the brand image
--keycloak-admin--brand--Height: var(--pf-global--spacer--xl); 
```

**Example**
```css
// Don’t increase specificity
// Don’t use pixel values
.keycloak-admin--manage-columns__modal .pf-c-dropdown {
   margin-bottom: 24px
}

// Do use a new class
// Do use a PatternFly global spacer variable
.keycloak-admin--manage-columns__dropdown {
   margin-bottom: var(--pf-global--spacer--xl);
}
```
### Using utility classes

Utility classes can be used to add specific styling to a component, such as margin-bottom or padding. However, their use should be limited to one-off styling needs.  

For example, instead of using the utility class for margin-right multiple times, we should define a new Keycloak admin console class that adds  *margin-right: var(--pf-global--spacer--sm); *
And in this example, the new class can set the color appropriately as well.

> **Using a utility class **

Reviewer: Something messed up in the formatting here - *s are showing up in the markdown - maybe an extra space?

Author: fixed had extra spaces.
